### DISCUSSION TITLE
端口转发的奇怪问题

### DISCUSSION THREAD

**Discussion by MasterOfStar** [MikeWang000000/Natter#56] [Q&A]
使用本人惯用的socat做端口转发的时候报错，日志如下
```
root@NEWIFI3(d-team_newifi-d2):/overlay/opt# python natter.py -t 2081
[INFO] - Getting STUN server IP...
Traceback (most recent call last):
  File "/overlay/opt/natter.py", line 583, in <module>
    main()
  File "/overlay/opt/natter.py", line 574, in main
    natter = Natter(src_host, src_port, test_http=test_http, log_level=log_level)
  File "/overlay/opt/natter.py", line 427, in __init__
    self.keep_alive_sock = self._init_keep_alive_sock()
  File "/overlay/opt/natter.py", line 436, in _init_keep_alive_sock
    s.bind((self.source_ip, self.source_port))
OSError: [Errno 125] Address in use
```
使用防火墙下的端口转发则没有这个问题。

是因为这两者的工作原理不同么？

（另外大佬有考虑写个openwrt的luci插件么？）

**Comment by MikeWang000000** [comment]
是的，socat 默认不允许端口重用。

可以尝试使用类似 `tcp-listen:2081,reuseaddr,reuseport,fork` 的参数，但不一定有用。

用户级的端口转发有以下缺点：
1. 较为耗费资源，如CPU使用率；
2. 目标端口服务获取到的IP地址永远为`127.0.0.1`；
3. 可能不支持端口重用，导致无法打洞。

因此，能使用防火墙的端口转发，就不用折腾其他转发了。

（暂时还没有写插件的打算，希望有luci大神帮个忙）

**Comment by peaceanddemocracy** [comment]
socat似乎在iptables的下一层，我一开始也是进入误区，测试发现应用层的转发似乎是不行的，必须在网络层直接转发端口。

**Comment by rankaiyx** [comment]
但是很多情况没有办法操作iptables  比如没有root的安卓设备   在这种情况下  集成一个应用级的端口转发功能 应该是很好的

另外 不知是feature 还是bug  似乎-t的http服务只能连接一次   比如用手机作为客户端，用手机流量连接好后  关闭流量 再开启流量  再访问  就无法成功

**Comment by rankaiyx** [comment]
实测了一下 配合socat 可以转发ssh  命令如下
先运行
socat TCP4-LISTEN:27022,reuseaddr,reuseport,fork TCP4:127.0.0.1:8022
再运行
python natter.py -v 27022
然后通过外网ip端口就可以访问到本地的8022的ssh了

环境 termux  on Android7.1  Linux 3.10.104
电信光猫桥接 路由器拨号 NAT1  DMZ到Android设备

**Comment by guaguaya66** [comment]
> 但是很多情况没有办法操作iptables 比如没有root的安卓设备 在这种情况下 集成一个应用级的端口转发功能 应该是很好的
> 
> 另外 不知是feature 还是bug 似乎-t的http服务只能连接一次 比如用手机作为客户端，用手机流量连接好后 关闭流量 再开启流量 再访问 就无法成功

我也发现了，还以为我打洞失败了

**Comment by MikeWang000000** [comment]
是bug    (｡ ́︿ ̀｡)
在修了在修了

目前还没有版本号，只能算是个玩具
